Feature/refactor commandhandler

mj610-kim opened the pull request:
With this PR applied, the input file and output file must be passed as arguments.

For example: 
![image](https://user-images.githubusercontent.com/101768406/159385705-d9e32255-752e-48a3-8fd2-09e6b6fd3dd0.png)

mj610-kim committed:
fix: remove arbitrary constants from CommandHandler

The program should run on any valid input and output file names. Remove
arbitrary constants for inputfile and outputFile in CommandHandler, and
update Main so that the passed arguments are used as file names.

Signed-off-by: 김명종 <mj610.kim@gmail.com>

mj610-kim committed:
refactor: clean up CommandHandler#run()

CommandHandler's run() method had a lot of nested code. This is cleaned
up by extracting out the runCore() method, while the exception handling
code remains in run().

Signed-off-by: 김명종 <mj610.kim@gmail.com>

mj610-kim committed:
refactor: add IS_DEBUG_MODE flag to CommandHandler

There is a lot of log printing in CommandHandler. During debugging, this
logging is very important. But in production code, this will slow down
performance. So this patch adds the IS_DEBUG_MODE flag to turn on/off
logging easily.

Signed-off-by: 김명종 <mj610.kim@gmail.com>

kwonhj0116 commented on EmployeeManagementSystem/src/main/java/com/cra08/excellentcode/CommandHandler.java:
```diff
@@ -19,84 +19,86 @@
 
-    private static final String INPUT_FILE = "input.txt";
-    private static final String OUTPUT_FILE = "output.txt";
+    private static final boolean IS_DEBUG_MODE = true;
```
디버깅모드로 printLog의 출력제어가 가능하니 불필요한 로그 출력이 줄어들겠네요!

taemin00 replied:
디버깅 모드 굿굿입니다!!

kwonhj0116 commented on EmployeeManagementSystem/src/main/java/com/cra08/excellentcode/CommandHandler.java:
```diff
@@ -19,84 +19,86 @@
-                }
-            }
-        } catch (Exception e) {
+            runCore(localFileReader, localFileWriter);
```
indent를 줄이기 위해 메소드 추출하여 리팩토링하신 아이디어 굳굳입니다!!

seongjin-Park210 reviewed (approved):
refactoring 및 기능 추가 확인했습니다. 드디어 실행 파일을 만들 준비가 다 된 것 같네요!! 
수고 많으셨습니다!

kwonhj0116 commented on EmployeeManagementSystem/src/main/java/com/cra08/excellentcode/CommandHandler.java:
```diff
@@ -19,84 +19,86 @@
-        System.out.println("cmd: " + cmd);
 
-        String result = null;
         switch (cmd) {
```
추후 명령어 추가의 가능성을 고려하여 명령어도 모듈화하면 어떨지 제안해봅니다..! 
이미 handle 메소드는 만들어두셨으니 interface만 추가하면 switch 문을 제거할 수 있을 것 같습니다!

mj610-kim replied:
네, 지금 enum 클래스에 abstract method 등을 구현하는 식으로 clean up을 구상중입니다. 조금 시간이 걸릴 것 같아서, 별도의 PR로 만들도록 하겠습니다. 

좋은 의견 감사합니다!

taemin00 commented on EmployeeManagementSystem/src/main/java/com/cra08/excellentcode/CommandHandler.java:
```diff
@@ -19,84 +19,86 @@
-    public void run() throws IOException {
-        InputReader localFileReader = new LocalFileReader(INPUT_FILE);
+    public void run() {
+        InputReader localFileReader = new LocalFileReader(inputFile);
         localFileReader.open();
 
-        OutputWriter localFileWriter = new LocalFileWriter(OUTPUT_FILE);
+        OutputWriter localFileWriter = new LocalFileWriter(outputFile);
```
inputFile과 outputFile에 경로 조작을 방지하기 위한 구문을 넣어주면 더 좋을 것 같습니다!

mj610-kim replied:
제가 말씀을 정확히 이해했는지는 모르겠습니다만, 추가 설명 드립니다!

1. `inputFile`, `outputFile` 멤버변수는 `final` 키워드로 지정했기 때문에 `CommandHandler`의 생성자에서 정해지고, 그 이후 수정은 없습니다!
2. `LocalFileWriter`, `LocalFileReader`에서 `open()`할 때 파일 열기를 시도하고, 실패할 경우 `false`를 리턴하게 됩니다. 

2번에서 추가적인 처리가 필요하다는 말씀이신가요?

taemin00 replied:
음 제가 말한 경우는 secure 코딩의 경로 조작입니다!
예를 들어 아래와 같이, output path에 악의적으로 프로그램이 돌아가는 PC의 local file path를 입력한다면,
local file이 덮어쓰기가 되어버릴 수가 있습니다. 이러한 동작을 사전에 방지 할 수 있도록, input_path나 output_path로 들어가는 String을 검증해야 할 것 같습니다!

![image](https://user-images.githubusercontent.com/101778501/159389324-1a8cf458-40ba-4e5d-a410-cd9181b48ad1.png)

mj610-kim replied:
오, 좋은 의견 감사합니다! 수정하겠습니다.

mj610-kim replied:
최신 PR 확인 부탁드립니다!

taemin00 commented on EmployeeManagementSystem/src/main/java/com/cra08/excellentcode/CommandHandler.java:
```diff
@@ -19,84 +19,86 @@
         String cmd = CommandParser.getCommand(input);
         List<String> employees = CommandParser.getColumnData(input);
-        System.out.println("employee: " + Arrays.toString(employees.toArray()));
+        printLog("employee: " + Arrays.toString(employees.toArray()));
```
print log 시 현재 어떤 커맨드를 수행하고 있는지도 추가 되면 좋을 것 같습니다!
예를들어 [ADD] 와 같은 식으로 붙여주면 더 좋을 것 같네요

mj610-kim replied:
리뷰 감사합니다! 최신 PR에 반영했습니다.

taemin00 reviewed (approved):
수정사항 확인했습니다~

mj610-kim committed:
feat: add feature to verify file validity

In the main method, a path to a system file or otherwise sensitive file
may be given as the arguments. This can be problematic for the program.
Apply secure coding principles and verify arguments before running
program core.

Signed-off-by: 김명종 <mj610.kim@gmail.com>
